packages/flutter_tools/lib/src/daemon.dart

```diff
@@ -170,6 +177,30 @@ class DaemonStreams {
     inputStream = DaemonInputStreamConverter(rawInputStream).convertedStream,
     _logger = logger;
 
+  /// Creates a [DaemonStreams] that uses stdin and stdout as the underlying streams.
+  DaemonStreams.fromStdio(Stdio stdio, { required Logger logger })
+    : this(stdio.stdin, stdio.stdout, logger: logger);
+
+  /// Creates a [DaemonStreams] that uses [Socket] as the underlying streams.
+  DaemonStreams.fromSocket(Socket socket, { required Logger logger })
+    : this(socket, socket, logger: logger);
+
+  /// Connects to a server and creates a [DaemonStreams] from the connection as the underlying streams.
+  factory DaemonStreams.connect(String host, int port, { required Logger logger }) {
+    final Future<Socket> socketFuture = Socket.connect(host, port);
+    final StreamController<List<int>> inputStreamController = StreamController<List<int>>();
+    final StreamController<List<int>> outputStreamController = StreamController<List<int>>();
+    socketFuture.then((Socket socket) {
+      inputStreamController.addStream(socket);
+      socket.addStream(outputStreamController.stream);
+    }).onError((Object error, StackTrace stackTrace) {
+      logger.printError('Socket error: $error');
```

Comment:
It looks like here you'd want to do an `inputStreamController.addError(e, s)`, and `close()` the `outputStreamController`? Then the error will be handled by the error handler on the `listen()` call on `inputStream`.

Reply:
Done